File: leet_code_algorithms/diameter_of_binary_tree/diameter_of_binary_tree.ts
export {};

class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;
  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }
}

/**
 * @param {TreeNode} root
 * @return {number}
 */
const diameterOfBinaryTree = (root: TreeNode | null) => {
  let maxDepth = 0;
  const dfs = (node: TreeNode | null) => {
    if (!node) {
      return -1;
    }
    const left = dfs(node.left);
    const right = dfs(node.right);
    maxDepth = Math.max(left + right + 2, maxDepth);
    return 1 + Math.max(left, right);
  };
  dfs(root);
  return maxDepth;
};

File: leet_code_algorithms/diameter_of_binary_tree/Diameter_of_binary_tree.java
package diameter_of_binary_tree;

public class Diameter_of_binary_tree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    private int max_depth;

    public int diameterOfBinaryTree(TreeNode root) {
        this.max_depth = 0;
        getMaxDepth(root);
        return this.max_depth;
    }

    public int getMaxDepth(TreeNode node) {
        if (node == null) {
            return -1;
        }
        int left = getMaxDepth(node.left);
        int right = getMaxDepth(node.right);
        this.max_depth = Math.max(left + right + 2, this.max_depth);
        return 1 + Math.max(left, right);
    }
}
